### PR TITLE
[SPARK-25602][SQL] SparkPlan.getByteArrayRdd should not consume the input when not necessary

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

In `SparkPlan.getByteArrayRdd`, we should only call `it.hasNext` when the limit is not hit, as `iter.hasNext` may produce one row and buffer it, and cause wrong metrics.

## How was this patch tested?

new tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @kiszk @viirya @mgaido91

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 1]
```diff
@@ -453,45 +453,89 @@ case class RangeExec(range: org.apache.spark.sql.catalyst.plans.logical.Range)
     val localIdx = ctx.freshName("localIdx")
     val localEnd = ctx.freshName("localEnd")
     val range = ctx.freshName("range")
-    val shouldStop = if (parent.needStopCheck) {
-      s"if (shouldStop()) { $number = $value + ${step}L; return; }"
+
+    val processingLoop = if (parent.needStopCheck) {
+      // TODO (cloud-fan): do we really need to do the stop check within batch?
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: This is the motivation of bringing the discussion at https://github.com/apache/spark/pull/10989#discussion_r221961271

If it's OK to not interrupt the loop and buffer result rows for join, I think it's also OK here.

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala [hunk 2]
```diff
@@ -453,45 +453,89 @@ case class RangeExec(range: org.apache.spark.sql.catalyst.plans.logical.Range)
...
+      |   long $range = $batchEnd - $nextIndex;
       |   if ($range != 0L) {
       |     int $localEnd = (int)($range / ${step}L);
-      |     for (int $localIdx = 0; $localIdx < $localEnd; $localIdx++) {
-      |       long $value = ((long)$localIdx * ${step}L) + $number;
-      |       ${consume(ctx, Seq(ev))}
-      |       $shouldStop
+      |     $processingLoop
+      |   } else {
+      |     long $nextBatchTodo;
```
Comment (mgaido91) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: why did you move these lines in the else?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 1 above)
Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: if we don't, then we would consume more rows than needed, don't we?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 3]
```diff
@@ -517,4 +517,93 @@ class SQLMetricsSuite extends SparkFunSuite with SQLMetricsTestUtils with Shared
...
+
+  test("SPARK-25602: range metrics can be wrong if the result rows are not fully consumed") {
+    val df = spark.range(0, 30, 1, 2).toDF().filter('id % 3 === 0)
+
+    withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true") {
+      df.collect()
+      df.queryExecution.executedPlan.foreach {
+        case w: WholeStageCodegenExec =>
+          w.child.foreach {
+            case f: FilterExec => assert(f.metrics("numOutputRows").value == 10L)
```
Comment (mgaido91) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: not a big issue, but if later we change things and these are not anymore here, we would not run the assert here. I would suggest to collect the `FilterExec` and the `RangeExec` and enforce that we collected 1 of both and then assert on them. What do you think?

Moreover, nit: would it be possible to dedup the code here? The tests are very similar with codegen on and off, only collecting the two exec nodes differs...

**Review by mgaido91 — COMMENTED** [review]
thanks for pinging me @cloud-fan.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: we just buffer more rows in `BufferRowIterator.currentRows`, it's only about performance IIUC.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: Now we don't do `return` when we need to stop the loop. Move these lines to else, so that we won't hit this code path when loop is interrupted.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 1 above)
Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: mmmh, but `localIdx` would become `localEnd` then, right? So the UTs you added would fail, or am I missing something?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 2 above)
Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: I see, but in this way we are looping 2 more times in the outer loop, because we either go in the if or in the else while previously we were doing both on the same iteration IIUC. I don't think it is a big issue but it may introduce a (very small probably) overhead compared to the previous case.

Since if IIUC in the first iteration we just go to the else branch now, since `batchEnd` is inited to `nextIndex`, do you think it is feasible to move this block before the inner loop? So we would solve both issues, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: I think that there is BroadcastHashJoin case doesn't mean it is generally ok to buffer more rows. If it is possible, we still should avoid it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 4]
```diff
@@ -517,4 +517,93 @@ class SQLMetricsSuite extends SparkFunSuite with SQLMetricsTestUtils with Shared
...
+            case r: RangeExec => assert(r.metrics("numOutputRows").value == 4L)
+            case _ =>
+          }
+
+        case _ =>
+      }
+    }
+
+    withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "false") {
+      // Top-most limit will only run the first task, so totally the Filter produces 2 rows, and
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: Does first task mean first partition?

**Comment by viirya** [comment]
retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: yes

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: > mmmh, but localIdx would become localEnd then, right? So the UTs you added would fail, or am I missing something?

You can pull my PR and try, the whole stage codegen is a little convoluted. `localIndex` is local and always starts with 0, `localEnd` is decided by the global `nextIndex`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala]: good idea!

**Comment by cloud-fan** [comment]
I simplified this PR to focus on `SparkPlan.getByteArrayRdd` only. Will submit PR to fix range later.

**Comment by gatorsmile** [comment]
retest this please

**Review by gatorsmile — COMMENTED** [review]
LGTM pending Jenkins

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 5]
```diff
@@ -517,4 +517,57 @@ class SQLMetricsSuite extends SparkFunSuite with SQLMetricsTestUtils with Shared
   test("writing data out metrics with dynamic partition: parquet") {
     testMetricsDynamicPartition("parquet", "parquet", "t1")
   }
+
+  test("SPARK-25602: SparkPlan.getByteArrayRdd should not consume the input when not necessary") {
+    def checkFilterAndRangeMetrics(
+        df: DataFrame,
+        filterNumOutputs: Int,
+        rangeNumOutputs: Int): Unit = {
+      var filter: FilterExec = null
```
Comment (mgaido91) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: what about something like this:
```
def collectExecNode[T](pf: PartialFunction[SparkPlan, T]): PartialFunction[SparkPlan, T] = {
        pf.orElse {
          case w: WholeStageCodegenExec =>
            w.child.collect(pf).head
        }
      }
      val range = df.queryExecution.executedPlan.collectFirst(
        collectExecNode { case r: RangeExec => r })
      val filter = df.queryExecution.executedPlan.collectFirst(
        collectExecNode { case f: FilterExec => f })
```

**Review by mgaido91 — COMMENTED** [review]
LGTM apart from one minor comment. Thanks

**Comment by viirya** [comment]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: In the future if we need to catch more nodes, we should abstract it. But for now it's only range and filter, I think it's ok.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.4!

**Comment by HyukjinKwon** [comment]
Let's say, this can be behaivour changes too since metrics are now changed. Should we update migration guide for safety?

**Comment by cloud-fan** [comment]
why do we need migration guide for bug fix?
